[ckaayy] iP

ckaayy opened the pull request:
### YapperBot

> "Your mind is for having ideas, not holding them" - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

YapperBot frees your mind of having to remember things you need to do. It's,

- Free 🤑 
- Easy to use 👍 
- User friendly! 💯 🥇 

All you need to do is,

1. download it from [here](https://github.com/ckaayy/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

**So, what are you waiting forrrrrr?**

Feature:

- [ ] Add todo tasks without deadline
- [ ] Add deadline tasks with deadline
- [ ] Add event tasks with duration
- [ ] Change task completion status
- [ ] Find and delete outdated tasks

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:


    public static void main(String[] args) {
        new YapperBot().run();
    }

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ckaayy committed:
Rename, Greet, Exit

ckaayy committed:
Added Echo feature

ckaayy committed:
Add, List

ckaayy committed:
Mark as Done

ckaayy committed:
ToDo, Event, Deadline

ckaayy committed:
Automated Text UI Testing

ckaayy committed:
Handle Errors

ckaayy committed:
Delete

ckaayy committed:
Use Enums

ckaayy committed:
fixed minor bugs in exception message

ckaayy committed:
Added functions to save task to a data file when tasks are added/ modified and load task from data file when YapperBot is run.

Each run of YapperBot does not retain information about previous runs and list of tasks are reset, making it difficult for future reference.

Let's
1. Store store the list of tasks to a text file in a intuitive format
2. Read from the text file to populate the arraylist at the start of the program so as to track past tasks

ckaayy committed:
Merge branch 'branch-Level-7'

ckaayy committed:
YapperBot can now read certain formats of input strings, interpret them as time, and output the time in another format

YapperBot needs to convert string input into meaningful time output for deadline and event tasks.

Let's,
1. Accept 2 format of input strings for deadline dates (yyyy-MM-dd or d/M/yyyy HHmm).
2. Accept strings in the format of "d/M/yyyy HHmm" for event tasks
3. Print out dates stored in the format of "MMM dd yyyy" or "MMM dd yyyy h:mm a"

ckaayy committed:
Merge branch 'branch-Level-8'

ckaayy committed:
Made the programme more object oriented.

Making the program more object-oriented offers several advantages that enhance the structure, maintainability, scalability, and readability of the code.

Lets,
1. Extract the code into 4 separate classes, Ui, Storage, Parser, TaskList. Each class deals with separate functionalities

ckaayy committed:
Merge branch 'branch-A-MoreOOP'

ckaayy committed:
Organised the classes into yapyap package.

Moving all the classes into a package has several advantages that contribute to better code organization, maintainability, and scalability of the Java application.

Let's
1. Add everything into a package named "yapyap"

ckaayy committed:
Merge branch 'branch-A-Packages'

ckaayy committed:
Merge branch 'add-gradle-support'

ckaayy committed:
Added unit tests.

Unit tests isolate a small unit of code to test its functionality, making sure there are no bugs and we are achieving the desired outcome with our implementation.

Let's,
1. Add unit tests to test the creation of todo, deadline and event tasks.
2. Add unit tests to test the marking and unmarking of all types of tasks.
3. Add unit tests to test the parser's parsing of user input.
4. Add unit tests to test the addition and deletion of task by the TaskList class.

ckaayy committed:
Merge branch 'branch-A-JUnit'

ckaayy committed:
Added Javadoc for all existing methods.

Javadocs provide clear and structured documentation that explains what each class, method, and field does, making it easier for other developers to understand the code quickly without having to dive into the implementation details.

Let's,
1. Add JavaDocs to all existing methods.

ckaayy committed:
Added the functionality of allowing users to find a task by searching for a keyword in the task description.

Allows users to search through the tasklist with ease, especially when the list gets too long.

Let's,
1. Add take user input and extract the search term
2. Make use of the search term to match with existing tasks in the task list to display tasks that contain the specified search term.

ckaayy committed:
Merge branch 'branch-A-JavaDoc'

ckaayy committed:
Merge branch 'branch-Level-9'

ckaayy committed:
Use CheckStyle to detect coding style violations

It is time consuming and error prone to manually look through code to identify style violations.

Let's,
1. Configure checkstyle.xml to define coding style to adhere to.
2. Use gradle to check for style errors within the project

ckaayy committed:
Merge branch 'branch-A-CheckStyle'

oneBoz commented on src/main/java/yapyap/Commands.java:
```diff
@@ -0,0 +1,15 @@
+package yapyap;
```
I like really like the name yapyap. But perhaps the files can be better separated into different packages

oneBoz commented on src/main/java/yapyap/Parser.java:
```diff
@@ -0,0 +1,241 @@
+            int index = Integer.parseInt(userInput.split(" ")[1]) - 1;
+            tasks.get(index).mark();
+            storage.saveTasks(tasks.getTasks());
+            System.out.println("________________________________");
```
I like how the print statements are so neat!

oneBoz commented on src/main/java/yapyap/Storage.java:
```diff
@@ -0,0 +1,93 @@
+                BufferedReader reader = Files.newBufferedReader(datafilePath);
+                String line;
+                while ((line = reader.readLine()) != null) {
+                    String[] taskInfo = line.split(" \\| ");
```
I like how the temporary variables are named nicely

oneBoz commented on src/main/java/yapyap/Task.java:
```diff
@@ -0,0 +1,72 @@
+ * Represents a task with a description and a completion status.
+ * This class serves as a base class for more specific types of tasks such as Todos, Deadlines, and Events.
+ */
+public class Task {
```
Perhaps a separate Todo class to handle Todo tasks, while Task is an abstract class?

oneBoz reviewed (commented):
Overall LGTM! The naming standards are well followed. Just some minor suggestions

ckaayy committed:
Add GUI for YapperBot

GUI makes the application more visually appealing for users.

Let's,
* make use of javaFX to create a GUI for YapperBot
* modify the format of response strings from YapperBot to suit the GUI

ckaayy committed:
Format dialog box

The size of the dialog box was too big.
Attempted to make the user icon circular but currently still working on it.

Let's decrease the width and height of the user icon and bot icon.

ckaayy committed:
Merge branch 'branch-A-Level-10'

ckaayy committed:
Add assertions

Assertions were not present previously

Add assertions to document important assumptions at various points of the code.

Adding these assertions can help catch errors early during development by making sure that certain conditions hold true, which are critical for the application to function correctly.

Let's,
* Add assertions to the most important files YapperBot.java, Parser.java, and MainWindow.java.

ckaayy committed:
Improve code quality

Refactor methods to reduce method complexity and highlight happy path.

Currently, some methods are too complicated and some happy paths are indented in an else block.

Refactor the methods by breaking them into smaller, more manageable pieces and bringing the happy path to the forefront.

This approach makes the code more readable and maintainable by clearly presenting the primary functionality of the method, while handling edge cases and exceptions separately.

ckaayy committed:
Merge pull request #1 from ckaayy/branch-A-CodeQuality

Improve code quality

ckaayy committed:
Refactor date parsing and formatting in Deadline class

The Deadline class has duplicated code for parsing and formatting dates in multiple methods.

Duplicated code makes the class harder to maintain and increases the risk of inconsistencies if the date format needs to be updated.

Extract the date parsing and formatting logic into separate methods and constants to centralize date handling.

Centralizing the date parsing and formatting improves readability, reduces redundancy, and makes the code easier to update and maintain.

ckaayy committed:
Merge pull request #2 from ckaayy/branch-A-Assertions

Add assertions

Assertions were not present previously

Add assertions to document important assumptions at various points of the code.

Adding these assertions can help catch errors early during development by making sure that certain conditions hold true, which are critical for the application to function correctly.

Let's,

Add assertions to the most important files YapperBot.java, Parser.java, and MainWindow.java.

ckaayy committed:
Add 'help' command to provide in-app guidance to users

YapperBot does not have a dedicated command for providing users with guidance.

Let's,
1. Add 'help' command to the Commands enum.
2. Implement the 'help' command logic in the Parser class to handle the display of help information.
3. Include a new method in Ui class to display the help message clearly.

The help command reduce learning curves for new users.

The help message includes details about all available commands and their expected formats.

ckaayy committed:
Add 'help' command to provide in-app guidance to users

YapperBot does not have a dedicated command for providing users instructions on how to use the bot, leading to potential confusion for new users.

Providing in-app help can improve user experience by offering accessible instructions.

Let’s,
1. Add 'help' command to the Commands enum.
2.  Implement the 'help' command logic in the Parser class to handle the display of help information.
3.  Include a new method in Ui class to display the help message clearly.
Merge branch 'branch-C-Help'

ckaayy committed:
Add 'help' command to provide in-app guidance to users

YapperBot does not have a dedicated command for providing users with guidance.

Let's,
1. Add 'help' command to the Commands enum.
2. Implement the 'help' command logic in the Parser class to handle the display of help information.
3. Include a new method in Ui class to display the help message clearly.

The help command reduce learning curves for new users.

The help message includes details about all available commands and their expected formats.

ckaayy committed:
Add welcome message

The application did not display a welcome message upon launching the GUI, which could lead to a lack of guidance for users on initial use.

Displaying a welcome message enhances user experience by providing an immediate indication that the application is ready and responsive.

Let's,
1. Add a displayWelcomeMessage method in the Ui class to show a welcome message when the application starts.
2. Integrated the welcome message display in the `MainWindow` initialization process.

By placing the welcome message in the `Ui` class, we maintain a clear separation of concerns, keeping UI-related logic centralized and reusable.

ckaayy committed:
Add welcome message

The application did not display a welcome message upon launching
the GUI, which could lead to a lack of guidance for users on
initial use.

Displaying a welcome message enhances user experience by
providing an immediate indication that the application
is ready and responsive.

Let's,
1. Add a displayWelcomeMessage method in the Ui class to show
a welcome message when the application starts.
2. Integrated the welcome message display in the `MainWindow`
initialization process.

By placing the welcome message in the `Ui` class, we maintain
a clear separation of concerns, keeping UI-related logic
centralized and reusable.
Merge branch 'branch-WelcomeMessage'

ckaayy committed:
Remove unused UI argument from parser method

Ui argument for executeCommandAndGetResponse is no longer needed.

Let's,
* remove the extra Ui passed to the method.

ckaayy committed:
Add unit tests for Storage class

The Storage class handles loading and saving tasks from/to a file, but
there are no tests to validate its functionality.

Lack of tests means potential bugs and incorrect behavior in file
handling could go unnoticed, leading to issues in task management.

Let's,
add comprehensive JUnit tests for the Storage class to verify loading
and saving tasks, including:
- Loading tasks from a valid file
- Handling non-existent files
- Throwing exceptions for invalid task types
- Saving tasks in the correct format

ckaayy committed:
Add unit tests for Storage class

The Storage class handles loading and saving tasks from/to a file, but
there are no tests to validate its functionality.

Lack of tests means potential bugs and incorrect behavior in file
handling could go unnoticed, leading to issues in task management.

Let's,
add comprehensive JUnit tests for the Storage class to verify loading
and saving tasks, including:
- Loading tasks from a valid file
- Handling non-existent files
- Throwing exceptions for invalid task types
- Saving tasks in the correct format

ckaayy committed:
Add comprehensive user guide for YapperBot

A user guide for YapperBot was missing, making it difficult for users to
understand the bot's features and commands.

Let's Add a detailed user guide that explains all the essential
commands, including task management, viewing, updating, and deleting
tasks, along with examples.

The user guide follows a simple and concise format to enhance usability
while providing necessary details for both new and experienced users.

ckaayy committed:
Add comprehensive user guide for YapperBot

A user guide for YapperBot was missing, making it difficult for users to
understand the bot's features and commands.

Let's Add a detailed user guide that explains all the essential
commands, including task management, viewing, updating, and deleting
tasks, along with examples.

The user guide follows a simple and concise format to enhance usability
while providing necessary details for both new and experienced users.

Merge branch 'branch-A-UserGuide'

ckaayy committed:
Add files via upload

ckaayy committed:
Delete docs/F2BEF4D1-59D8-4B9C-ADB4-BB664B43A6E1.jpeg

ckaayy committed:
Add Ui.png to docs folder

ckaayy committed:
Fix bugs in parser class

FindCommand: function in parser class return the wrong string when there
are no matching tasks found.

Let's return the no match string immediately without appending
additional statements.

ckaayy committed:
Update README.md

ckaayy committed:
Fix bugs when user input is null

Parser class: Handles scenario when user input is null.

ckaayy committed:
Fix compatibility issues on linux and windows OS

Build.gradle: changed mainclass to a class that do not extend
application
